### PR TITLE
PARQUET-2305 Allow Parquet to Proto conversion even though Target Schema has less fields

### PR DESCRIPTION
**Pull request by tddfan**
If Parquet has any field which has been removed from the schema and Parquet to Proto conversion happens, it errors out due to Unknown fields. There could be some scenarios that we want to still convert PARQUET into the target proto schema object which has lesser fields. In this change using the ProtoParquetReader and specifying "ignoreUnknownFields" as an argument, the conversion can still be done which would ignore fields it can't convert and not error out.

Similar functionality exist in https://github.com/protocolbuffers/protobuf/blob/main/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java with field "ignoringUnknownFields"

Make sure you have checked all steps below.

Tests
[ SS] My PR adds more cases to the unit test ProtoSchemaConverterTest to cover for ignoreUnknownFields:

### CODE REVIEW

**Comment by wgtmac** [comment]
@tddfan Thanks for the fix. Would you please create a JIRA issue and link it to the title?

**Comment by tddfan** [comment]
> @tddfan Thanks for the fix. Would you please create a JIRA issue and link it to the title?

Thanks for the guidance. I have created jira and put this PR title. Please let me know if anything else is needed.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 1]
```diff
@@ -86,32 +89,71 @@ class ProtoMessageConverter extends GroupConverter {
     this.conf = conf;
     this.parent = pvc;
     this.extraMetadata = extraMetadata;
-    int parquetFieldIndex = 1;
+    boolean ignoreUnknownFields = conf.getBoolean("IGNORE_UNKNOWN_FIELDS", false);
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:92]: Seems line 92-94 can be moved under line 98?

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 2]
```diff
@@ -86,32 +89,71 @@ class ProtoMessageConverter extends GroupConverter {
...
+
+    myBuilder = builder;
 
     if (pvc == null) {
       throw new IllegalStateException("Missing parent value container");
     }
 
-    myBuilder = builder;
+    if(builder == null && ignoreUnknownFields) {
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:100]:
```suggestion
    if (builder == null && ignoreUnknownFields) {
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 3]
```diff
@@ -86,32 +89,71 @@ class ProtoMessageConverter extends GroupConverter {
...
-      Descriptors.FieldDescriptor protoField = protoDescriptor.findFieldByName(parquetField.getName());
+      int parquetFieldIndex = 0;
+      Descriptors.Descriptor protoDescriptor =  builder.getDescriptorForType();
 
-      if (protoField == null) {
-        String description = "Scheme mismatch \n\"" + parquetField + "\"" +
-                "\n proto descriptor:\n" + protoDescriptor.toProto();
-        throw new IncompatibleSchemaModificationException("Cant find \"" + parquetField.getName() + "\" " + description);
-      }
+      for (Type parquetField : parquetSchema.getFields()) {
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:109]: The indentation in the for loop looks weird to me. Could you please format it a little bit?

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 4]
```diff
@@ -86,32 +89,71 @@ class ProtoMessageConverter extends GroupConverter {
...
 
+  private Converter dummyScalarConverter(ParentValueContainer pvc,
+                                         Type parquetField, Configuration conf,
+                                         Map<String, String> extraMetadata) {
+
+    if(parquetField.isPrimitive()) {
+      PrimitiveType primitiveType = parquetField.asPrimitiveType();
+      PrimitiveType.PrimitiveTypeName primitiveTypeName = primitiveType.getPrimitiveTypeName();
+      switch (primitiveTypeName) {
+        case BINARY: return new ProtoStringConverter(pvc);
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:142]: CMIW, do we miss some cases? It looks shorter than here: https://github.com/apache/parquet-mr/blob/master/parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java#L176-L184

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 5]
```diff
@@ -124,13 +166,15 @@ public void start() {
 
   @Override
   public void end() {
-    parent.add(myBuilder.build());
-    myBuilder.clear();
+    if(myBuilder != null) {
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:169]:
```suggestion
    if (myBuilder != null) {
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 6]
```diff
@@ -124,13 +166,15 @@ public void start() {
...
+      parent.add(myBuilder.build());
+      myBuilder.clear();
+    }
   }
 
   protected Converter newMessageConverter(final Message.Builder parentBuilder, final Descriptors.FieldDescriptor fieldDescriptor, Type parquetType) {
 
-    boolean isRepeated = fieldDescriptor.isRepeated();
+    boolean isRepeated = fieldDescriptor==null ? false : fieldDescriptor.isRepeated();
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:177]:
```suggestion
    boolean isRepeated = fieldDescriptor != null && fieldDescriptor.isRepeated();
```

File: parquet-protobuf/src/test/java/org/apache/parquet/proto/TestUtils.java [hunk 7]
```diff
@@ -195,6 +196,18 @@ public static <T extends MessageOrBuilder> List<T> readMessages(Path file, Class
...
+  /**
+   * Read messages from given file into the expected proto class.
+   * @param file
+   * @param messageClass
+   * @param <T>
+   * @return List of protobuf messages for the given type.
+   */
+  public static <T extends MessageOrBuilder> List<T> readMessages(Path file, Class<T> messageClass) throws IOException {
+    return readMessages(file, messageClass, false);
+
```
Comment (wgtmac) [inline parquet-protobuf/src/test/java/org/apache/parquet/proto/TestUtils.java:208]: Please remove this blank line.

File: parquet-protobuf/src/test/java/org/apache/parquet/proto/ProtoSchemaEvolutionTest.java [hunk 8]
```diff
@@ -65,4 +66,68 @@ public void testEnumSchemaWriteV1ReadV2() throws IOException {
     assertEquals(messagesV2.size(), 1);
     assertSame(messagesV2.get(0).getOptionalLabelNumberPair(), TestProto3SchemaV2.MessageSchema.LabelNumberPair.SECOND);
   }
+
+  /**
```
Comment (wgtmac) [inline parquet-protobuf/src/test/java/org/apache/parquet/proto/ProtoSchemaEvolutionTest.java:70]: Thanks for adding the test!

**Review by wgtmac — COMMENTED** [review]
Thanks for the fix! I have left some comments.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java [hunk 9]
```diff
@@ -37,11 +37,17 @@
   public static <T> ParquetReader.Builder<T> builder(Path file) {
     return new ProtoParquetReader.Builder<T>(file);
   }
-
+  public static <T> ParquetReader.Builder<T> builder(Path file, boolean ignoreUnknownFields) {
```
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java:43]: Done.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java [hunk 10]
```diff
@@ -71,6 +77,13 @@ protected Builder(InputFile file) {
       super(file);
     }
 
+    private Builder setIgnoreUnknownFields(boolean ignoreUnknownFields) {
+      if(ignoreUnknownFields) {
```
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java:85]: Change made.

File: parquet-protobuf/src/test/java/org/apache/parquet/proto/ProtoSchemaEvolutionTest.java (hunk 8 above)
Comment (tddfan) [reply parquet-protobuf/src/test/java/org/apache/parquet/proto/ProtoSchemaEvolutionTest.java:70]: Thanks :)

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 11]
```diff
@@ -18,9 +18,7 @@
  */
 package org.apache.parquet.proto;
 
-import com.google.protobuf.ByteString;
-import com.google.protobuf.Descriptors;
-import com.google.protobuf.Message;
+import com.google.protobuf.*;
```
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java]: Resolved this.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java [hunk 12]
```diff
@@ -71,6 +77,13 @@ protected Builder(InputFile file) {
...
+        this.set("IGNORE_UNKNOWN_FIELDS", "TRUE");
```
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoParquetReader.java]: Good point. Implemented this.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java (hunk 5 above)
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java]: Done

File: parquet-protobuf/src/test/java/org/apache/parquet/proto/TestUtils.java (hunk 7 above)
Comment (tddfan) [reply parquet-protobuf/src/test/java/org/apache/parquet/proto/TestUtils.java]: Done

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java (hunk 2 above)
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java]: Change done.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 13]
```diff
@@ -31,9 +29,7 @@
 import org.apache.parquet.io.api.Converter;
 import org.apache.parquet.io.api.GroupConverter;
 import org.apache.parquet.io.api.PrimitiveConverter;
-import org.apache.parquet.schema.GroupType;
-import org.apache.parquet.schema.IncompatibleSchemaModificationException;
-import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.*;
```
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java]: Resolved.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java [hunk 14]
```diff
@@ -26,6 +26,7 @@ public final class ProtoConstants {
   public static final String METADATA_ENUM_PREFIX = "parquet.proto.enum.";
   public static final String METADATA_ENUM_KEY_VALUE_SEPARATOR = ":";
   public static final String METADATA_ENUM_ITEM_SEPARATOR = ",";
+  public static final String IGNORE_UNKNOWN_FIELDS = "IGNORE_UNKNOWN_FIELDS";
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java:29]: In addition, could you also add some comment to describe its purpose?

**Review by wgtmac — COMMENTED** [review]
Thanks for the change! Left some cosmetic comments.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java (hunk 1 above)
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java]: I think the whole check can for pvv==null can be moved up. I have done that. Please let me know if you disagree

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java [hunk 15]
```diff
@@ -26,6 +26,7 @@ public final class ProtoConstants {
   public static final String METADATA_ENUM_PREFIX = "parquet.proto.enum.";
   public static final String METADATA_ENUM_KEY_VALUE_SEPARATOR = ":";
   public static final String METADATA_ENUM_ITEM_SEPARATOR = ",";
+  public static final String CONFIG_IGNORE_UNKNOWN_FIELDS = "parquet.proto.ignore.unknwon.fields";
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java:29]:
```suggestion
  public static final String CONFIG_IGNORE_UNKNOWN_FIELDS = "parquet.proto.ignore.unknown.fields";
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 16]
```diff
@@ -31,10 +32,12 @@
 import org.apache.parquet.io.api.Converter;
 import org.apache.parquet.io.api.GroupConverter;
 import org.apache.parquet.io.api.PrimitiveConverter;
+
 import org.apache.parquet.schema.GroupType;
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:36]:
```suggestion
import org.apache.parquet.io.api.PrimitiveConverter;
import org.apache.parquet.schema.GroupType;
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 17]
```diff
@@ -80,35 +91,77 @@ class ProtoMessageConverter extends GroupConverter {
...
+      int parquetFieldIndex = 0;
+      Descriptors.Descriptor protoDescriptor =  builder.getDescriptorForType();
 
-      if (protoField == null) {
-        String description = "Scheme mismatch \n\"" + parquetField + "\"" +
-                "\n proto descriptor:\n" + protoDescriptor.toProto();
-        throw new IncompatibleSchemaModificationException("Cant find \"" + parquetField.getName() + "\" " + description);
+      for (Type parquetField : parquetSchema.getFields()) {
+
+        Descriptors.FieldDescriptor protoField = protoDescriptor.findFieldByName(parquetField.getName());
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:118]:
```suggestion
      for (Type parquetField : parquetSchema.getFields()) {
        Descriptors.FieldDescriptor protoField = protoDescriptor.findFieldByName(parquetField.getName());
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java [hunk 18]
```diff
@@ -80,35 +91,77 @@ class ProtoMessageConverter extends GroupConverter {
...
+  }
 
-      converters[parquetFieldIndex - 1] = newMessageConverter(myBuilder, protoField, parquetField);
 
-      parquetFieldIndex++;
+  private Converter dummyScalarConverter(ParentValueContainer pvc,
+                                         Type parquetField, Configuration conf,
+                                         Map<String, String> extraMetadata) {
+
+    if(parquetField.isPrimitive()) {
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoMessageConverter.java:147]:
```suggestion
    if (parquetField.isPrimitive()) {
```

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java (hunk 15 above)
Comment (wgtmac) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java:29]: Could you please add a comment for `CONFIG_IGNORE_UNKNOWN_FIELDS`? Like what we did for `CONFIG_ACCEPT_UNKNOWN_ENUM`.

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tddfan) [reply parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java]: Done

File: parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java [hunk 19]
```diff
@@ -26,6 +26,12 @@ public final class ProtoConstants {
   public static final String METADATA_ENUM_PREFIX = "parquet.proto.enum.";
   public static final String METADATA_ENUM_KEY_VALUE_SEPARATOR = ":";
   public static final String METADATA_ENUM_ITEM_SEPARATOR = ",";
+  /**
+   * Configuration flag to ignore the unknowns fields during conversion of Parquet to Proto
```
Comment (wgtmac) [inline parquet-protobuf/src/main/java/org/apache/parquet/proto/ProtoConstants.java:30]:
```suggestion
   * Configuration flag to ignore the unknown fields during conversion of Parquet to Proto
```

**Review by wgtmac — APPROVED** [review]
LGTM. Thanks @tddfan!

**Pull request merged**